astropy/utils/data.py

```diff
@@ -1313,43 +1415,309 @@ def _get_download_cache_locs():
+                # could include threading.current_thread().ident as well
+                pid_based_random = hash(str(os.getpid()))/2.**sys.hash_info.width
+                dt = 0.05*(1+pid_based_random)
+                time.sleep(dt)
+                waited += dt
+            else:
+                # Never did get the lock.
+                raise RuntimeError(msg)
 
-        except OSError:
-            time.sleep(1)
+        yield
+
+    finally:
+        if os.path.exists(lockdir):
+            if os.path.isdir(lockdir):
+                # if the pid file is present, be sure to remove it
+                if os.path.exists(pidfn):
+                    os.remove(pidfn)
+                os.rmdir(lockdir)
+            else:
+                msg = 'Error releasing lock. "{}" exists but is not a directory.'
+                raise RuntimeError(msg.format(lockdir))
         else:
-            return
-    msg = ("Unable to acquire lock for cache directory ({0} exists). "
-           "You may need to delete the lock if the python interpreter wasn't "
-           "shut down properly.")
-    raise RuntimeError(msg.format(lockdir))
+            # Just in case we were called from _clear_download_cache
+            # or something went wrong before creating the directory
+            pass
 
 
-def _release_download_cache_lock():
-    lockdir = os.path.join(_get_download_cache_locs()[0], 'lock')
+class ReadOnlyDict(dict):
+    def __setitem__(self, key, value):
+        raise TypeError("This object is read-only.")
+
+
+nothing = ReadOnlyDict()  # might as well share.
+
 
-    if os.path.isdir(lockdir):
-        # if the pid file is present, be sure to remove it
-        pidfn = os.path.join(lockdir, 'pid')
-        if os.path.exists(pidfn):
-            os.remove(pidfn)
-        os.rmdir(lockdir)
+@contextlib.contextmanager
+def _cache(write=False):
+    """Download cache context manager.
+
+    Yields a pair consisting of the download cache directory and a dict-like
+    object that maps downloaded URLs to the filenames that contain their
+    contents; these files reside in the download cache directory.
+
+    If writing is requested, this holds the lock and yields a modifiable
+    dict-like object (actually a shelve object from the shelve module).  If
+    there is something wrong with the cache setup, an appropriate exception
+    will be raised.
+
+    If reading is requested, the lock will be briefly acquired, the URL map
+    will be copied into a read-only dict-like object which is yielded, and the
+    lock will be released. If some problem occurs and the cache is inaccessible
+    or non-functional, a CacheMissingWarning will be emitted and this context
+    manager will yield an empty dict-like object and None as the download
+    directory.
+
+    The files in the download cache directory are named according to a
+    cryptographic hash of their contents (currently MD5, so hackers can cause
```

Comment:
As already mentioned in https://github.com/astropy/astropy/issues/9303#issuecomment-536620522 that's just incorrect. The least of the worries if the download-site is hacked is that the checksums match (because the hash is just for duplicate-matching).